src/Framework/Migrations/Controllers/BatchMigrationRunner.php

```diff
@@ -0,0 +1,214 @@
+{
+    /**
+     * @var BatchMigration
+     */
+    private $migration;
+
+    /**
+     * @var ActionScheduler_Store
+     */
+    private $actionSchedulerStore;
+
+    /**
+     * @unreleased
+     */
+    public function __construct(BatchMigration $migration)
+    {
+        $this->migration = $migration;
+        $this->actionSchedulerStore = ActionScheduler_Store::instance();
+    }
+
+    /**
+     * Run batch migration
+     *
+     * @unreleased
+     *
+     * @throws Exception
+     */
+    public function run(): string
+    {
+        $this->registerAction();
+
+        $actions = $this->actionSchedulerStore->query_actions([
+            'group' => $this->getGroup(),
+            'per_page' => 0,
+        ]);
+
+        if ( ! $actions) {
+            $itemsCount = $this->migration->getItemsCount();
+
+            // Bailout if there are no items to process
+            if ( ! $itemsCount) {
+                return MigrationLogStatus::SUCCESS;
+            }
+
+            $current = 0;
+            $batches = ceil($itemsCount / $this->migration->getBatchSize());
+
+            // Register migration action for each batch
+            for ($i = 0; $i < $batches; $i++) {
+                if ($items = $this->migration->getBatchItemsAfter($current)) {
+                    [$firstId, $lastId] = $items;
+                    $current = $lastId;
+                    as_enqueue_async_action($this->getHook(), [$firstId, $lastId], $this->getGroup());
+                }
+            }
+
+            return MigrationLogStatus::RUNNING;
+        }
+
+        $pendingActions = (int)$this->actionSchedulerStore->query_actions([
+            'group' => $this->getGroup(),
+            'status' => [
+                ActionScheduler_Store::STATUS_RUNNING,
+                ActionScheduler_Store::STATUS_PENDING,
+            ],
+        ], 'count');
+
+        if ($pendingActions) {
+            return MigrationLogStatus::RUNNING;
+        }
+
+        $failedActions = (int)$this->actionSchedulerStore->query_actions([
+            'group' => $this->getGroup(),
+            'status' => [
+                ActionScheduler_Store::STATUS_FAILED,
+                ActionScheduler_Store::STATUS_CANCELED,
+            ],
+        ], 'count');
+
+        if ($failedActions) {
+            return MigrationLogStatus::INCOMPLETE;
+        }
+
+        // Run the last check
+        if ($this->migrationHasIncomingData()) {
+            return MigrationLogStatus::RUNNING;
+        }
+
+        // If everything went well, delete scheduled actions
+        foreach ($actions as $actionId) {
+            $this->actionSchedulerStore->delete_action($actionId);
+        }
+
+        return MigrationLogStatus::SUCCESS;
+    }
+
+    /**
+     * @unreleased
+     */
+    public function getHook(): string
+    {
+        return 'givewp-batch-' . $this->migration::id();
+    }
+
+    /**
+     * @unreleased
+     */
+    public function getGroup(): string
```

Comment:
Same. Looks like this can be private.

Reply:
Resolved https://github.com/impress-org/givewp/pull/7755/commits/39639508d9282b6b0f385c9947666f9946cb7bfd